File: benches/python/quantum_benchmark.py
import numpy as np
import time
import csv

# Pauli matrices
PAULI_X = np.array([[0, 1], [1, 0]], dtype=complex)
PAULI_Y = np.array([[0, -1j], [1j, 0]], dtype=complex)
PAULI_Z = np.array([[1, 0], [0, -1]], dtype=complex)
HADAMARD = (1/np.sqrt(2)) * np.array([[1, 1], [1, -1]], dtype=complex)

def apply_gate(state, gate):
    """Apply gate to quantum state"""
    return gate @ state

def fidelity(state1, state2):
    """Calculate fidelity between two states"""
    inner = np.vdot(state1, state2)
    return np.abs(inner) ** 2

def benchmark_pauli_x(size):
    state = np.array([1, 0], dtype=complex)
    
    start = time.perf_counter()
    for _ in range(size):
        state = apply_gate(state, PAULI_X)
    end = time.perf_counter()
    
    return end - start

def benchmark_hadamard(size):
    state = np.array([1, 0], dtype=complex)
    
    start = time.perf_counter()
    for _ in range(size):
        state = apply_gate(state, HADAMARD)
    end = time.perf_counter()
    
    return end - start

def benchmark_fidelity_calc(size):
    bell1 = (1/np.sqrt(2)) * np.array([1, 0, 0, 1], dtype=complex)
    bell2 = (1/np.sqrt(2)) * np.array([1, 0, 0, 1], dtype=complex)
    
    start = time.perf_counter()
    for _ in range(size):
        f = fidelity(bell1, bell2)
    end = time.perf_counter()
    
    return end - start

def benchmark_bell_creation(size):
    start = time.perf_counter()
    for _ in range(size):
        bell = (1/np.sqrt(2)) * np.array([1, 0, 0, 1], dtype=complex)
    end = time.perf_counter()
    
    return end - start

def run_benchmarks():
    sizes = [1_000, 10_000, 100_000]
    results = []
    
    print("Python Quantum Operations Benchmark")
    print("=" * 50)
    
    benchmarks = [
        ("Pauli-X", benchmark_pauli_x),
        ("Hadamard", benchmark_hadamard),
        ("Fidelity", benchmark_fidelity_calc),
        ("Bell State", benchmark_bell_creation),
    ]
    
    for name, func in benchmarks:
        print(f"\n{name}:")
        for size in sizes:
            times = []
            for _ in range(5):
                times.append(func(size))
            
            avg_time = sum(times) / len(times)
            print(f"  {size:,}: {avg_time*1000:.3f} ms")
            
            results.append({
                "language": "Python",
                "benchmark": name,
                "size": size,
                "time_ms": avg_time * 1000
            })
    
    # Save to CSV
    with open("benches/results/python_quantum_results.csv", "w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=["language", "benchmark", "size", "time_ms"])
        writer.writeheader()
        writer.writerows(results)
    
    print("\nSaved to benches/results/python_quantum_results.csv")

if __name__ == "__main__":
    run_benchmarks()

File: benches/quantum_benchmark.rs
use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
use std::hint::black_box;
use QComNetSim::quantum::gates::{hadamard, pauli_x, pauli_y, pauli_z};
use QComNetSim::quantum::{Qubit, TwoQubitState};

fn benchmark_single_qubit_gates(c: &mut Criterion) {
    let mut group = c.benchmark_group("Single Qubit Gates");

    for size in [1_000, 10_000, 100_000].iter() {
        group.bench_with_input(BenchmarkId::new("Pauli-X", size), size, |b, &size| {
            b.iter(|| {
                let mut qubit = Qubit::new_zero();
                for _ in 0..size {
                    pauli_x(&mut qubit);
                }
                black_box(qubit);
            });
        });

        group.bench_with_input(BenchmarkId::new("Hadamard", size), size, |b, &size| {
            b.iter(|| {
                let mut qubit = Qubit::new_zero();
                for _ in 0..size {
                    hadamard(&mut qubit);
                }
                black_box(qubit);
            });
        });
    }

    group.finish();
}

fn benchmark_fidelity_calculation(c: &mut Criterion) {
    let mut group = c.benchmark_group("Fidelity Calculation");

    let bell1 = TwoQubitState::new_bell_phi_plus();
    let bell2 = TwoQubitState::new_bell_phi_plus();

    for size in [1_000, 10_000, 100_000].iter() {
        group.bench_with_input(BenchmarkId::new("Fidelity", size), size, |b, &size| {
            b.iter(|| {
                for _ in 0..size {
                    black_box(bell1.fidelity(&bell2));
                }
            });
        });
    }

    group.finish();
}

fn benchmark_state_creation(c: &mut Criterion) {
    let mut group = c.benchmark_group("State Creation");

    for size in [1_000, 10_000, 100_000].iter() {
        group.bench_with_input(BenchmarkId::new("Bell State", size), size, |b, &size| {
            b.iter(|| {
                for _ in 0..size {
                    black_box(TwoQubitState::new_bell_phi_plus());
                }
            });
        });
    }

    group.finish();
}

criterion_group!(
    benches,
    benchmark_single_qubit_gates,
    benchmark_fidelity_calculation,
    benchmark_state_creation
);
criterion_main!(benches);
